File: src/main/java/io/github/alexiscomete/lapinousecond/worlds/Place.java
package io.github.alexiscomete.lapinousecond.worlds;

import io.github.alexiscomete.lapinousecond.Main;
import io.github.alexiscomete.lapinousecond.save.*;
import org.javacord.api.entity.message.embed.EmbedBuilder;

import java.awt.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Optional;

public class Place extends CacheGetSet {
    private Long serverID;
    private ServerBot serverBot;
    private World world;
    private Integer x;
    private Integer y;
    private long[] connections;

    public Place() {
        super(SaveLocation.generateUniqueID(), Tables.PLACES.getTable());
        HashMap<String, String> h = new HashMap<>();
        h.put("id", String.valueOf(getID()));
        Main.getSaveManager().insert("places", h);
    }

    public Place(long id) {
        super(id, Tables.PLACES.getTable());
    }

    public Optional<ServerBot> getServerBot() {
        return Optional.ofNullable(serverBot);
    }

    public Optional<Long> getServerID() {
        return Optional.ofNullable(serverID);
    }

    public void setServerBot(ServerBot serverBot) {
        this.serverBot = serverBot;
    }

    public void setServerID(Long serverID) {
        this.serverID = serverID;
    }

    public World getWorld() {
        return world;
    }

    public void setWorld(World world) {
        this.world = world;
    }

    public Optional<Integer> getX() {
        return Optional.ofNullable(x);
    }

    public void setX(Integer x) {
        this.x = x;
    }

    public Optional<Integer> getY() {
        return Optional.ofNullable(y);
    }

    public void setY(Integer y) {
        this.y = y;
    }

    public long getID() {
        return id;
    }

    public long[] getConnections() {
        return connections;
    }

    public void setConnections(long[] connections) {
        this.connections = connections;
    }

    public Place setAndGet(String row, String value) {
        super.set(row, value);
        return this;
    }

    public EmbedBuilder getPlaceEmbed() {
        return new EmbedBuilder()
                .setAuthor(String.valueOf(id))
                .setTitle(getString("name"))
                .setColor(Color.green)
                .setDescription(getString("descr"))
                .addField("World", getString("world"), true)
                .addField("Type", getString("type"), true);
    }

    public static ArrayList<Place> toPlaces(String places) {
        String[] str = places.split(";");
        ArrayList<Place> places1 = new ArrayList<>();
        for (String s :
                str) {
            places1.add(new Place(Long.parseLong(s)));
        }
        return places1;
    }
}

File: src/main/java/io/github/alexiscomete/lapinousecond/worlds/Place.kt
package io.github.alexiscomete.lapinousecond.worlds

import io.github.alexiscomete.lapinousecond.entity.concrete.resources.Resource
import io.github.alexiscomete.lapinousecond.entity.entities.Owner
import io.github.alexiscomete.lapinousecond.useful.managesave.*
import java.sql.SQLException
import java.util.*

val PLACES = Table("places")
val places = CacheCustom(PLACES) { id: Long -> Place(id) }

open class Place(
    id: Long = run {
        val tempID = generateUniqueID()
        val h = HashMap<String, String>()
        h["id"] = tempID.toString()
        saveManager.insert("places", h)
        tempID
    }
) : CacheGetSet(id, PLACES), Owner {

    open fun getX(): Optional<Int> {
        return Optional.ofNullable(this["x"].toInt())
    }

    open fun getY(): Optional<Int> {
        return Optional.ofNullable(this["y"].toInt())
    }

    override val ownerType: String = ""
    override val ownerString: String = ""
    override fun addMoney(amount: Double) {
        TODO("Not yet implemented")
    }

    override fun getMoney(): Double {
        TODO("Not yet implemented")
    }

    override fun removeMoney(amount: Double) {
        TODO("Not yet implemented")
    }

    override fun addResource(resource: Resource, amount: Double) {
        TODO("Not yet implemented")
    }

    override fun getResource(resource: Resource): Double {
        TODO("Not yet implemented")
    }

    override fun removeResource(resource: Resource, amount: Double) {
        TODO("Not yet implemented")
    }

    override fun hasResource(resource: Resource, amount: Double): Boolean {
        TODO("Not yet implemented")
    }

    override fun hasMoney(amount: Double): Boolean {
        TODO("Not yet implemented")
    }

}

fun toPlaces(places: String): ArrayList<Place> {
    val str = places.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
    val places1 = ArrayList<Place>()
    for (s in str) {
        try {
            places1.add(Place(s.toLong()))
        } catch (ignore: NumberFormatException) {
        }
    }
    return places1
}

fun getPlacesWithWorld(world: String): ArrayList<Place> {
    val resultSet = saveManager.executeQuery("SELECT * FROM places WHERE world = '$world'", true)
    val places = ArrayList<Place>()
    try {

        //long id = resultSet.getLong("id"); places.add(new Place(id));
        while (resultSet!!.next()) {
            places.add(Place(resultSet.getLong("id")))
        }
    } catch (e: SQLException) {
        e.printStackTrace()
    }
    return places
}